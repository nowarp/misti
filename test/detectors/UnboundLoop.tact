fun test1(): Int {
    let a: Int = 20; // read-only variable
    while (a > 10) { // ERROR: a is not changing within the loop
    }
    return a;
}

fun test2(): Int {
    let a: Int = 20; // read-only variable
    do  { // ERROR: a is not changing within the loop
    } until (a > 10);
    return a;
}

fun test3(): Int {
    let a: Int = 20;
    while (a > 10)  { // OK: a is changing within the loop
        a += 1;
    }
    return a;
}

fun test4(): Int {
  let a: Int = 10;
  repeat (a) {} // no warning, since the repeat is bounded by definition
  return a;
}

fun test5() {
    let i: Int = 0;
    while (i > 0) {} // WARNING: unbound loop
}

fun test6() {
    let i: Int = 0;
    while (10 < i) {} // WARNING: unbound loop
}


fun test7() {
    let i: Bool = false;
    while (i) {} // WARNING: unbound loop
}

fun test8() {
    let i: Bool = true;
    while (!i) {} // WARNING: unbound loop
}


fun fn(a: Int) {
}

fun test9(): Int {
    let a: Int = 20;
    while (a > 10)  { // OK: a is using within the loop
        fn(a);
    }
    return a;
}


struct S {
    a: Int;
}

contract Test {
    receive("test") {
        let s: S = S {
            a: 1,
        };
        let i: Int = 0;
        while (i < s.a) { // OK: No warning
            i += 1;
        }
    }
}