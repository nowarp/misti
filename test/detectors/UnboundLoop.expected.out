[HIGH] UnboundLoop: Found unbounded loop
test/detectors/UnboundLoop.tact:3:12:
  2 |     let a: Int = 20; // read-only variable
> 3 |     while (a > 10) { // ERROR: a is not changing within the loop
                 ^
  4 |     }
The condition variable doesn't change inside the loop
Help: Consider changing the variable inside the loop to ensure it terminates
See: https://nowarp.io/tools/misti/docs/detectors/UnboundLoop

[HIGH] UnboundLoop: Found unbounded loop
test/detectors/UnboundLoop.tact:11:14:
  10 |     do  { // ERROR: a is not changing within the loop
> 11 |     } until (a > 10);
                    ^
  12 |     return a;
The condition variable doesn't change inside the loop
Help: Consider changing the variable inside the loop to ensure it terminates
See: https://nowarp.io/tools/misti/docs/detectors/UnboundLoop

[HIGH] UnboundLoop: Found unbounded loop
test/detectors/UnboundLoop.tact:31:12:
  30 |     let i: Int = 0;
> 31 |     while (i > 0) {} // WARNING: unbound loop
                  ^
  32 | }
The condition variable doesn't change inside the loop
Help: Consider changing the variable inside the loop to ensure it terminates
See: https://nowarp.io/tools/misti/docs/detectors/UnboundLoop

[HIGH] UnboundLoop: Found unbounded loop
test/detectors/UnboundLoop.tact:36:17:
  35 |     let i: Int = 0;
> 36 |     while (10 < i) {} // WARNING: unbound loop
                       ^
  37 | }
The condition variable doesn't change inside the loop
Help: Consider changing the variable inside the loop to ensure it terminates
See: https://nowarp.io/tools/misti/docs/detectors/UnboundLoop

[HIGH] UnboundLoop: Found unbounded loop
test/detectors/UnboundLoop.tact:42:12:
  41 |     let i: Bool = false;
> 42 |     while (i) {} // WARNING: unbound loop
                  ^
  43 | }
The condition variable doesn't change inside the loop
Help: Consider changing the variable inside the loop to ensure it terminates
See: https://nowarp.io/tools/misti/docs/detectors/UnboundLoop

[HIGH] UnboundLoop: Found unbounded loop
test/detectors/UnboundLoop.tact:47:13:
  46 |     let i: Bool = true;
> 47 |     while (!i) {} // WARNING: unbound loop
                   ^
  48 | }
The condition variable doesn't change inside the loop
Help: Consider changing the variable inside the loop to ensure it terminates
See: https://nowarp.io/tools/misti/docs/detectors/UnboundLoop